src/Hangfire.Mongo/Signal/Mongo/MongoSignal.cs

```diff
@@ -0,0 +1,159 @@
+        }
+
+        public void Set(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("Name should not be empty", nameof(name));
+            }
+
+            var update = Builders<SignalDto>.Update
+                .Set(s => s.Signaled, true)
+                .Set(s => s.Name, name)
+                .CurrentDate(s => s.TimeStamp);
+            _signal.UpdateOne(_ => false, update, new UpdateOptions { IsUpsert = true });
+
+            Logger.Debug($@"*** SET: {name} ({Thread.CurrentThread.ManagedThreadId})");
+        }
+
+        public void Wait(string name)
+        {
+            Wait(name, TimeSpan.MaxValue);
+        }
+
+        public void Wait(string name, TimeSpan timeout)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("Name should not be empty", nameof(name));
+            }
+
+            var e = EventWaitHandles.GetOrAdd(name, n => new EventWaitHandle(false, EventResetMode.AutoReset));
+            Logger.Debug($@">>> WAIT: {name} ({Thread.CurrentThread.ManagedThreadId})");
+            if (timeout == TimeSpan.MaxValue)
+            {
+                e.WaitOne();
+            }
+            else
+            {
+                e.WaitOne(timeout);
+            }
+            Logger.Debug($@"<<< WAIT: {name} ({Thread.CurrentThread.ManagedThreadId})");
+        }
+
+        public void Wait(string name, CancellationToken cancellationToken)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException("Name should not be empty", nameof(name));
+            }
+
+            Logger.Debug($@">>> WAIT: {name} ({Thread.CurrentThread.ManagedThreadId})");
+            Task.Run(() => Wait(name, TimeSpan.MaxValue)).Wait(cancellationToken);
+            Logger.Debug($@"<<< WAIT: {name} ({Thread.CurrentThread.ManagedThreadId})");
+        }
+
+        public void Wait(string[] names, CancellationToken cancellationToken)
+        {
+            if (names == null)
+            {
+                throw new ArgumentNullException(nameof(names));
+            }
+            if (names.Length == 0 || names.Any(string.IsNullOrEmpty))
+            {
+                throw new ArgumentException("Name should not be empty", nameof(names));
+            }
+
+            Logger.Debug($@">>> WAIT: {string.Join(",", names)} ({Thread.CurrentThread.ManagedThreadId})");
+            Task.WaitAny(names.Select(n => Task.Run(() => Wait(n, TimeSpan.MaxValue))).ToArray(), cancellationToken);
```

Comment:
Maybe create multiple entries using the same cancellationTokens.WaitHandle? and avoid the Task.Run

Reply:
I'll look into that...

Reply:
Using WaitHandle.WaitAny instead - much better solution.